src/dhtproto/client/internal/NodeHashRanges.d

```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
 
     /***************************************************************************
 
+        Helper encapsulating the node-selection logic required by requests that
+        get a single record from the DHT. Namely:
+
+            During a data redistribution, more than one node may be responsible
+            for a given key. In this case, the node that was most recently
+            reported as being responsible for the key is queried first, followed
+            by others (in order) until the record is located, an error occurs,
+            or no node has the record.
+
+        TODO: test the logic for retrying the request on other nodes which
+        previously covered the hash. This will require a full neo implementation
+        of the Redistribute request. See
+        https://github.com/sociomantic/dhtnode/issues/624
+
+        Params:
+            h = hash to query
+            node_hash_ranges = buffer to receive hash range information of
+                nodes which cover the specified hash
+            use_node = delegate to call to gain access to a request-on-conn to
+                communicate with the selected node
+            get = delegate to communicate with the selected node (over the
+                request-on-conn provided by use_node). Should return true if the
+                get succeeded or false to try the next node
+
+        Returns:
+            true if there are no nodes covering the specified hash
+
+    ***************************************************************************/
+
+    public bool getFromNode ( hash_t h, ref NodeHashRange[] node_hash_ranges,
+        UseNodeDg use_node,
+        bool delegate ( RequestOnConn.EventDispatcher ) get )
+    {
+        auto nodes = this.getNodesForHash(h, node_hash_ranges);
+        if ( nodes.length == 0 )
+            return true;
+
+        foreach_reverse ( node_hash_range; nodes )
```

Comment:
This `foreach_reverse` indicates that the ordering is somehow important. Maybe it should be added to the method's documentation for the parameter `getNodesForHash`.

Reply:
It already is described...?

>During a data redistribution, more than one node may be responsible
for a given key. In this case, the node that was most recently
reported as being responsible for the key is queried first, followed
by others (in order) until the record is located, an error occurs,
or no node has the record.

Reply:
Oh I see what you mean. `getNodesForHash` doesn't mention how the returned list is ordered.

Reply:
Weird, it appears it's actually unsorted! (I think this is what the TODO above is referring to ;)